Migrate to xunit.v3

Youssef1313 opened the pull request:
## Description

Re-creating the PR after some work on Arcade is done to support xunit.v3 extensions.

The number of jobs you have is **too large** to have a complete comparison. But taking few random samples:

Two runs used for comparison:

- Before: https://github.com/Youssef1313/aspire/actions/runs/14146248192/job/39633895044?pr=4
- After: https://github.com/Youssef1313/aspire/actions/runs/14146226732/job/39633854038

| Job and step                                                                                          | Before | After | Number of tests before | Number of tests after |
| ----------------------------------------------------------------------------------------------------- | ------ | ----- | ---------------------- | --------------------- |
| Setup for tests (Linux) - enumerate tests                                                             | 36s    | 34s   | 54                     | 54                    |
| Setup for tests (Windows) - enumerate tests                                                           | 2m 19s | 2m 8s | 33                     | 33                    |
| Integrations Linux (Azure.AI.OpenAI) / Azure.AI.OpenAI - run tests                                    | 14s    | 28s   | 113                    | 113                   |
| Integrations Linux (Azure.Data.Tables) / Azure.Data.Tables - run tests                                | 20s    | 21s   | 41                     | 41                    |
| Integrations Linux (Azure.Messaging.EventHubs) / Azure.Messaging.EventHubs - run tests                | 4s     | 5s    | 242                    | 242                   |
| Integrations Linux (Azure.Messaging.ServiceBus) / Azure.Messaging.ServiceBus - run tests              | 27s    | 28s   | 76                     | 76                    |
| Integrations Linux (Azure.Messaging.WebPubSub) / Azure.Messaging.WebPubSub - run tests                | 34s    | 41s   | 48                     | 48                    |
| Integrations Linux (Azure.Npgsql) / Azure.Npgsql - run tests                                          | 13s    | 17s   | 55                     | 55                    |
| EndToEnd - Run nuget dependent tests                                                                  | 52s    | 41s   | 4                      | 4                     |

NOTE: I'm comparing a single run before vs a single run after. There might already be high variance as every job is running really small number of tests. So numbers are unlikely to be meaningful here.

**NOTE 2**: There are 3 test failures that need to be looked at. The failures are similar so the root cause is the same. It might be related to how the test assembly is loaded causing working directory to be different or something.

Fixes # (issue)

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Is this introducing a breaking change?
      - [ ] Yes
        - Link to aspire-docs issue (please use this [`breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)):
      - [ ] No
        - Link to aspire-docs issue (please use this [`doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)):
  - [ ] No

Youssef1313 committed:
Update XUnit extensions from Arcade

Youssef1313 committed:
Update packages and Arcade test runner

Youssef1313 committed:
Move custom attributes to xunit.v3

Youssef1313 committed:
Fix usings

Youssef1313 committed:
Move from ConditionalFact/ConditionalTheory to Assert.Skip

Youssef1313 committed:
Suppress xUnit1030

Youssef1313 committed:
Remove unnecessary using

Youssef1313 committed:
Few fixes

Youssef1313 committed:
Exe

Youssef1313 committed:
Param doc

Youssef1313 committed:
Add OutputType

Youssef1313 committed:
Adjust for helix

Youssef1313 committed:
Produce binlog

Youssef1313 committed:
Specify DOTNET_ROOT for executables to find the correct runtime

Youssef1313 committed:
Fix build errors

danmoseley commented:
14->28s would be problematic if it's real. Could you maybe do repeat measurement to check whether it's normal variation?

Youssef1313 commented:
@danmoseley I just looked at another "before" build (https://github.com/dotnet/aspire/actions/runs/14140547808/job/39621122075?pr=8392) and that's 26s.

So looks like there is large variance already between runs that wouldn't make measurements here easy/useful. This is probably normal because the number of tests being run in each job is too small already.

Youssef1313 commented:
@danmoseley For the change in one of the test results, I detailed the reason in the commit message (https://github.com/dotnet/aspire/commit/bb67af0d7dbcbb746eb7f23a001adc8cbfa529fe)

Youssef1313 committed:
Fix failing test

This test used to pass with VSTest and xunit v2 because the tests were run under testhost.exe.

So, DistributedApplicationOptions.Assembly was pointing to testhost. Then ResolveProjectDirectory wouldn't correctly find AppHostProjectPath. In DistributedApplicationBuilder constructor, we were falling back to _innerBuilder.Environment.ContentRootPath because ProjectDirectory is null.

With xunit.v3 *or* MTP, generally when the test app is executed as normal executable, we have the right assembly and we are able to resolve project directory correctly

Youssef1313 committed:
Fix for Arcade

Youssef1313 commented:
The test `EndToEndLoggingTest` look like it's incorrectly authored. It's currently written as follows

```csharp
        logger.LoggedMessage = () =>
        {
            // wait for at least 2 logs to be written
            if (logger.Logs.Count >= 2)
            {
                tsc.SetResult();
            }
        };
```

If we got a third log message, that will crash because the Task already completed.

Youssef1313 committed:
Fix test

eerhardt commented on tests/Aspire.Azure.Search.Documents.Tests/Aspire.Azure.Search.Documents.Tests.csproj:
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>$(DefaultTargetFramework)</TargetFramework>
+    <OutputType>Exe</OutputType>
```
Can this be refactored into test\Directory.Build.targets or similar? that way we don't need it on every test.

eerhardt replied:
> Directory.Build.targets is too late to set OutputType

It shouldn't be. If you set it before you import whatever .targets file that reads it.

Youssef1313 replied:
> You could set it automatically in Directory.Build.props if the project name ends with .Tests.csproj

Yeah I think good idea.

---

> > Directory.Build.targets is too late to set OutputType
>
> It shouldn't be. If you set it before you import whatever .targets file that reads it.

I don't remember all the details, but the .NET SDK itself imports Directory.Build.targets after some other stuff that the SDK imports internally (I can't recall the names exactly). But basically the OutputType is read by SDK in some places before Directory.Build.targets is imported.

In VSTest, this is already problematic as VSTest will try to set OutputType in package .targets. SDK already tried to workaround it https://github.com/dotnet/sdk/blob/532dc9df641e03ebf3faf54ddab1018cb096f556/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Common.targets#L30-L34 but even that workaround isn't working reliably in some niche cases.

Youssef1313 replied:
I moved it to Directory.Build.props. Let's see if everything is still green.

RussKie replied:
(A bit of OT) I faced similar issues in https://github.com/dotnet/extensions/pull/5640#discussion_r1853289921. There are some extensibility points available, which could provide an alternative solution, e.g.:
https://github.com/dotnet/extensions/blob/c43462a31323873b8e4329f6b8f86cc46da6c8ee/Directory.Build.props#L109-L121

Youssef1313 replied:
Oh, nice. I think the `BeforeMicrosoftNETSdkTargets` could work. I'll keep it as Directory.Build.props though as it seems simpler and is working well so far.

Youssef1313 committed:
Move to props

radical commented:
We should run a azdo build also.

Youssef1313 commented:
@radical Could you do please?

radical commented:
> @radical Could you do please?

https://dev.azure.com/dnceng-public/public/_build/results?buildId=999739&view=results

radical commented:
There are a bunch of failures. The pipeline tests, for example, are failing with:
```
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v3.0.2+dd36e86129 (64-bit .NET 8.0.13)
You must install or update .NET to run this application.
App: /mnt/vss/_work/1/s/artifacts/bin/Aspire.Cli.Tests/Release/net8.0/Aspire.Cli.Tests
Architecture: x64
Framework: 'Microsoft.NETCore.App', version '8.0.0' (x64)
.NET location: /usr/share/dotnet
The following frameworks were found:
  6.0.36 at [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Learn about framework resolution:
https://aka.ms/dotnet/app-launch-failed
To install missing framework, download:
https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=8.0.0&arch=x64&rid=ubuntu.20.04-x64
[xUnit.net 00:00:00.18] Aspire.Cli.Tests: Catastrophic failure: System.InvalidOperationException: Test process terminated unexpectedly.
   at Xunit.Runner.v3.Xunit3..ctor(XunitProjectAssembly projectAssembly, ISourceInformationProvider sourceInformationProvider, IMessageSink diagnosticMessageSink, ITestProcessLauncher testProcessLauncher)
   at Xunit.Runner.v3.Xunit3.ForDiscoveryAndExecution(XunitProjectAssembly projectAssembly, ISourceInformationProvider sourceInformationProvider, IMessageSink diagnosticMessageSink, ITestProcessLauncher testProcessLauncher)
   at Xunit.XunitFrontController.Create(XunitProjectAssembly projectAssembly, ISourceInformationProvider sourceInformationProvider, IMessageSink diagnosticMessageSink, ITestProcessLauncher testProcessLauncher)
   at Xunit.Runner.VisualStudio.VsTestRunner.RunTestsInAssembly(IRunContext runContext, IFrameworkHandle frameworkHandle, LoggerHelper logger, TestPlatformContext testPlatformContext, RunSettings runSettings, IMessageSink reporterMessageHandler, AssemblyRunInfo runInfo)
   at Xunit.Runner.VisualStudio.VsTestRunner.RunTestsInAssembly(IRunContext runContext, IFrameworkHandle frameworkHandle, LoggerHelper logger, TestPlatformContext testPlatformContext, RunSettings runSettings, IMessageSink reporterMessageHandler, AssemblyRunInfo runInfo)
No test matches the given testcase filter `category!=failing` in /mnt/vss/_work/1/s/artifacts/bin/Aspire.Cli.Tests/Release/net8.0/Aspire.Cli.Tests.dll
```

Youssef1313 committed:
Specify DOTNET_ROOT

Youssef1313 committed:
Fix ExtractTestClassNames for local SDK usage as well

radical commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -92,6 +92,7 @@ steps:
         DCP_DIAGNOSTICS_LOG_LEVEL: debug
         DCP_DIAGNOSTICS_LOG_FOLDER: $(Build.ArtifactStagingDirectory)/artifacts/log/dcp
         DCP_PRESERVE_EXECUTABLE_LOGS: 1
+        DOTNET_ROOT: $(Build.SourcesDirectory)/.dotnet
```
Why is this needed? The build, and dotnet scripts should be able to pick the correct dotnet to use. `Build.SourcesDirectory)/.dotnet` will not be available if the required version is already installed system-wide.

Youssef1313 replied:
@radical The tests are run now as a separate executable which won't be aware of the correct .NET installation.

radical replied:
Would this be a problem when this path does not exist? What about local runs?

Youssef1313 replied:
This is CI-only configuration. It doesn't affect local runs

radical commented:
> Thanks @radical. Could you re-run again please with the last change I did?

https://dev.azure.com/dnceng-public/public/_build/results?buildId=999842&view=results

radical commented:
These need to be checked before merging:

- [ ] number of tests run in GH actions hasn't changed
- [ ] number of tests run on Azdo pipeline (including helix) hasn't changed
- [ ] Azdo can still ingest the test results
- [ ] trx files are being generated, and picked up for artifacts in GH actions, and azdo

Youssef1313 commented:
That failure may be Helix-related. @ViktorHofer @akoeplinger Do you know please?

radical commented:
I can take a look at the failures when I get a chance.

RussKie commented on tests/Aspire.Components.Common.Tests/ConformanceTests.cs:
```diff
@@ -88,23 +87,23 @@ protected virtual void DisableRetries(TOptions options) { }
         if (typeof(TOptions) == typeof(object))
         {
-            throw new SkipTestException("Not implemented yet");
+            Assert.Skip("Not implemented yet");
```
Oh, this is nice! Finally, we'd be able to use it.

RussKie commented on tests/Aspire.Dashboard.Components.Tests/Shared/IntegrationTestHelpers.cs:
```diff
@@ -3,6 +3,7 @@
 
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Logging.Testing;
+using Xunit;
```
Instead of adding an explicit using, please add an implicit one instead (like you did in #8310).

Youssef1313 replied:
@RussKie I think that's better done as a separate change as the analyzer for unused usings is enabled in this repo and there are many existing `using Xunit;`. The goal of #8310 was to minimize the diff when working on this PR.

RussKie commented on tests/Aspire.TestUtilities/RequiresDockerAttribute.cs:
```diff
@@ -29,4 +28,14 @@ public RequiresDockerAttribute(string? reason = null)
         Reason = reason;
     }
+
+    public IReadOnlyCollection<KeyValuePair<string, string>> GetTraits()
+    {
+        if (!IsSupported)
+        {
+            return [new KeyValuePair<string, string>(XunitConstants.Category, "failing")];
+        }
+
+        return [];
+    }
```
So, there are no more discoverers, which is cool.

How does this work though? Where does `IsSupported` come from and how is it set?

Youssef1313 replied:
IsSupported was, and is still, defined in the attribute class

https://github.com/dotnet/aspire/blob/ad75817e69a1494b6c878ece595b0dfc42801e5f/tests/Aspire.TestUtilities/RequiresDockerAttribute.cs#L23-L26

Just that traits calculation is moved from a special trait discoverer to the attribute class itself.

RussKie commented on tests/Directory.Build.props:
```diff
@@ -15,6 +15,8 @@
     <!-- This is useful for local runs -->
     <VSTestResultsDirectory>$(ArtifactsTestResultsDir)</VSTestResultsDirectory>
+
+    <OutputType Condition="$(MSBuildProjectName.EndsWith('.Tests'))">Exe</OutputType>
```
Any reason not go all the way and use `<Project Sdk="MSTest.Sdk/x.y.z">` which, essentially, means the tests are run as exe?

Youssef1313 replied:
This repo is xUnit, not MSTest.

Youssef1313 committed:
Set executable bit

Youssef1313 commented:
I pushed one more change to see if it resolves the permission denied error on Helix. @radical if you can run an AzDO build please :) Thanks!

radical commented:
Validation build - https://dev.azure.com/dnceng-public/public/_build/results?buildId=1000431&view=results

Youssef1313 committed:
Set xunit.analyzers version

Youssef1313 commented:
@radical I did one more small fix

Youssef1313 committed:
Disable analyzer for now

Youssef1313 committed:
Disable analyzer for now

Youssef1313 committed:
Move to Directory.Build.props as NoWarn

Youssef1313 committed:
Adjust

Youssef1313 committed:
Disable more warnings

Youssef1313 committed:
Suppress more warnings

Youssef1313 committed:
Add back to props

Youssef1313 committed:
Merge branch 'main' into xunit.v3-round-2

Youssef1313 commented:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=1000823&view=results

It should get green.

Youssef1313 commented:
Last build on main https://dev.azure.com/dnceng-public/public/_build/results?buildId=1000565&view=ms.vss-test-web.build-test-results-tab:

AzDO build on current PR: https://dev.azure.com/dnceng-public/public/_build/results?buildId=1000823&view=ms.vss-test-web.build-test-results-tab


Number of tests in both cases is 9001 and all TRXes are there.

Youssef1313 commented:
@radical @danmoseley @joperezr Hopefully this is good to merge now?

radical committed:
Merge branch 'main' into xunit.v3-round-2

radical committed:
fix merge

radical committed:
fix merge

Youssef1313 committed:
Merge branch 'main' into xunit.v3-round-2

radical committed:
Merge branch 'main' into xunit.v3-round-2

radical committed:
fix merge

radical commented on Directory.Build.props:
```diff
@@ -33,6 +33,11 @@
     <DashboardPublishedArtifactsOutputDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'DashboardArtifacts', '$(Configuration)'))</DashboardPublishedArtifactsOutputDir>
     <WorkloadsPackageSource>$(ArtifactsShippingPackagesDir)</WorkloadsPackageSource>
+
+    <!-- xUnit2031: Do not use Where clause with Assert.Single -->
```
Why are we disabling this one?

Youssef1313 replied:
@radical I wanted to avoid extra changes in this PR to keep it minimal. But I think it can be addressed later. I see you are already tracking it now in the follow-up issue.

radical commented on Directory.Build.props:
```diff
@@ -33,6 +33,11 @@
     <WorkloadsPackageSource>$(ArtifactsShippingPackagesDir)</WorkloadsPackageSource>
+
+    <!-- xUnit2031: Do not use Where clause with Assert.Single -->
+    <!-- xUnit1051: Calls to methods which accept CancellationToken should use TestContext.Current.CancellationToken to allow test cancellation to be more responsive. -->
```
Could you please open an issue to track this as a TODO, and reference that here?

radical commented on tests/Aspire.Templates.Tests/AppHostTemplateTests.cs:
```diff
@@ -1,6 +1,9 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
+// TODO: Fix this.
+#pragma warning disable xUnit1051 // Calls to methods which accept CancellationToken should use TestContext.Current.CancellationToken to allow test cancellation to be more responsive.
```
Disable this in the `Aspire.Templates.Tests.csproj` instead of adding to all the test files.

radical replied:
I opened https://github.com/dotnet/aspire/issues/8488 and pushed the change. Please feel free to discuss if you don't agree with the change.

radical committed:
Merge remote-tracking branch 'origin/main' into xunit.v3-round-2

# Conflicts:
#	tests/Aspire.Azure.Npgsql.EntityFrameworkCore.PostgreSQL.Tests/EnrichNpgsqlTests.cs

radical committed:
Remove unused properties for helix run

radical committed:
Disable xunit1051 from the project file for Playground, and Templates tests

radical committed:
add reference to issue

radical commented:
@danmoseley @eerhardt @joperezr @RussKie Any objections, or concerns about this?

radical committed:
Merge remote-tracking branch 'origin/main' into xunit.v3-round-2

radical committed:
Update QuarantinedTestAttribute for xunit v3

radical commented:
I have merged main, and fixed the new QuarantineTestAttribute . cc @RussKie

radical committed:
More fixes for merging main

Youssef1313 committed:
Adjust for QuarantinedTest

radical committed:
More adjustments for merging main

RussKie commented on tests/Aspire.Playground.Tests/QuarantinedTestAttribute.cs:
```diff
@@ -53,4 +51,7 @@ public QuarantinedTestAttribute(string? reason = null)
     }
+
+    public IReadOnlyCollection<KeyValuePair<string, string>> GetTraits()
+        => [new KeyValuePair<string, string>("quarantined", "true")];
```
There's one more 
https://github.com/dotnet/aspire/blob/3477fc1fec3fbc145883ff309cff89f6a072b16f/tests/Aspire.TestUtilities/QuarantinedTestAttribute.cs#L39-L54

RussKie replied:
Oh, it's been updated in e36881f96b75707a6a006e3704dc1ca8e14f9b59.

RussKie commented on eng/Xunit3/Xunit3.targets:
```diff
@@ -0,0 +1,18 @@
+<Project>
```
@Youssef1313 is there a reason this isn't done at the Arcade level?

Youssef1313 replied:
WIP: https://github.com/dotnet/arcade/pull/15668

Youssef1313 replied:
But you will need to wait until you are using latest Arcade 10 anyways

RussKie replied:
Awesome!

RussKie reviewed (commented):
Overall 👍

radical commented:
Validation run - https://dev.azure.com/dnceng-public/public/_build/results?buildId=1002032&view=results

radical reviewed (approved):
Thank you for working on this! Great contribution :) LGTM 👍

radical commented:
But do wait for the azdo build to be green too, before merging.

Youssef1313 commented:
Thanks @radical. I'll merge once all is green (including AzDO build) then.
